modules/11_subclustering/subclustering.qmd

```diff
@@ -0,0 +1,676 @@
+
+# USER INPUT: 
+# - Specify which barcodes of the default assay to remove
+bcs = SeuratObject::Cells(sc[[default_assay]])
+barcode_metadata = sc[[]][bcs, ]
+
+barcodes_to_remove = barcode_metadata %>%
+  dplyr::filter(seurat_clusters == "1") %>%
+  rownames()
+
+# Filter barcodes:
+# a) barcodes that are present only in this assay, will be removed from the whole Seurat object
+# b) barcodes that are present in other assays, will be removed from this assay only
+barcodes_other_assays = purrr::map(setdiff(SeuratObject::Assays(sc), default_assay), function(a) return(SeuratObject::Cells(sc[[a]]))) %>% 
+  purrr::flatten_chr()
+
+# Remove from assay barcodes that are still present in other assays
+barcodes_to_remove_assay = intersect(barcodes_to_remove, barcodes_other_assays)
+barcodes_to_keep = setdiff(SeuratObject::Cells(sc[[default_assay]]), barcodes_to_remove_assay)
+sc[[default_assay]] = subset(sc[[default_assay]], cells=barcodes_to_keep)
+
+# Remove from Seurat object barcodes that are not present in other assays
+barcodes_to_remove_sc = setdiff(barcodes_to_remove, barcodes_other_assays)
+if (length(barcodes_to_remove_assay) > 0) {
+  barcodes_sc = purrr::map(SeuratObject::Assays(sc), function(a) return(SeuratObject::Cells(sc[[a]]))) %>% purrr::flatten_chr() %>% unique()
+  barcodes_sc = setdiff(barcodes_sc, barcodes_to_remove_sc)
+  sc = subset(sc, cells=barcodes_sc)
+}
+
+# Indicate that barcodes were removed so that on-disk matrices (if used) are updated
+barcodes_removed = TRUE
+
+# Print information
+CalloutBox(x=FormatString("Removed {length(barcodes_to_remove)} barcodes from the analysis.", quote=FALSE), type="note")
+```
+
+```{r}
+#| label: fig-subclustering_remove_barcodes_2
+#| fig-cap: 'Clusters after removing barcodes'
+#| eval: false
+
+# Plot a UMAP
+DimPlot(sc, label=TRUE) + 
+  AddPlotStyle()
+```
+
+-->
+
+<!--
+## Merge/rename cluster
+
+USER_INPUT:
+
+- Remove the HTML comments
+- Set eval in the chunk options to true
+- Add explanation why clusters are merged/renamed here
+
+```{r}
+#| label: subclustering_merge_rename_cluster_1
+#| tbl-cap: 'Clusters after merging and/or renaming'
+#| eval: false
+
+# Use this chunk to merge or rename clusters in the Seurat object
+
+# Get barcodes for default assay
+bcs = SeuratObject::Cells(sc[[default_assay]])
+
+# USER_INPUT: 
+# - Set cluster column and what clusters to merge/rename into a new column 'merged_clusters'
+cluster_column = "seurat_clusters"
+
+# Always keep: 
+# - Clusters that are not merged or re-named remain the same as before
+# - Barcodes that are not part of the default assay will be set to NA
+sc$merged_clusters = ifelse(SeuratObject::Cells(sc) %in% bcs, 
+                            sc[[cluster_column, drop=TRUE]] %>% as.character(), 
+                            NA)
+
+# Specify what to merge/rename
+n = sc[[]] %>% 
+  dplyr::filter(merged_clusters %in% c("1", "2", "3")) %>%
```

Comment:
I don't understand :) - we can also talk in person.

Reply:
see above